src/ciphers/caesar.rs

```diff
@@ -1,43 +1,117 @@
+    if !(0..=25).contains(&rotation) {
+        return Err(ERROR_MESSAGE);
+    }
+
+    let result = text
         .chars()
         .map(|c| {
             if c.is_ascii_alphabetic() {
-                let first = if c.is_ascii_lowercase() { b'a' } else { b'A' };
-                // modulo the distance to keep character range
-                (first + (c as u8 + shift - first) % 26) as char
+                shift_char(c, rotation)
             } else {
                 c
             }
         })
-        .collect()
+        .collect();
+
+    Ok(result)
+}
+
+/// Shifts a single ASCII alphabetic character by a specified number of positions in the alphabet.
+///
+/// # Arguments
+///
+/// * `c` - The ASCII alphabetic character to be shifted.
+/// * `rotation` - The number of positions to shift the character. Should be within the range (0, 25).
+///
+/// # Returns
+///
+/// Returns the shifted ASCII alphabetic character.
+fn shift_char(c: char, rotation: isize) -> char {
+    let first = if c.is_ascii_lowercase() { b'a' } else { b'A' };
+    let rotation = rotation as u8; // Safe cast as rotation is within (0, 25)
+
+    (((c as u8 - first) + rotation) % 26 + first) as char
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    #[test]
-    fn empty() {
-        assert_eq!(caesar("", 13), "");
-    }
+    macro_rules! test_caesar {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (text, rotation, expected) = $test_case;
+
+                    match rotation {
+                        0..=25 => {
+                            // Test forward rotation
+                            match caesar(&text, rotation) {
+                                Ok(result) => assert_eq!(result, expected),
+                                Err(e) => panic!("Unexpected error: {}", e),
+                            }
 
-    #[test]
-    fn caesar_rot_13() {
-        assert_eq!(caesar("rust", 13), "ehfg");
+                            // Test backward rotation
+                            let backward_rotation = if rotation == 0 { 0 } else { 26 - rotation };
+                            match caesar(&expected, backward_rotation) {
+                                Ok(result) => assert_eq!(result, text),
+                                Err(e) => assert_eq!(e, ERROR_MESSAGE),
+                            }
+                        },
+                        _ => {
+                            // If rotation is invalid, test that the function returns the specific error message
+                            match caesar(&text, rotation) {
+                                Ok(_) => panic!("Expected an error but got an Ok result"),
+                                Err(e) => assert_eq!(e, ERROR_MESSAGE),
+                            }
+                        }
+                    }
+                }
+            )*
```

Comment:
This is way to complex. Please split it into two tests: one checking the _happy path_ (something like [suggested here](https://github.com/TheAlgorithms/Rust/pull/720#discussion_r1606048554)) and the second, with inputs, for which `caesar` returns an error.

Reply:
Created the following macro_rules: `test_caesar_happy_path` and `test_caesar_error_cases`.

I also updated the tests to match with the new rules.